crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -243,105 +245,107 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
-
         // add account changeset changes
         for (block_number, account_before) in account_changeset.into_iter().rev() {
             let AccountBeforeTx { info: old_info, address } = account_before;
-            match state.entry(address) {
-                hash_map::Entry::Vacant(entry) => {
-                    let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((old_info, new_info, HashMap::new()));
-                }
-                hash_map::Entry::Occupied(mut entry) => {
-                    // overwrite old account state.
-                    entry.get_mut().0 = old_info;
-                }
-            }
+            let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
+            bundle_builder = bundle_builder
+                .state_original_account_info(address, into_revm_acc(old_info.unwrap_or_default()));
+            bundle_builder = bundle_builder
+                .state_present_account_info(address, into_revm_acc(new_info.unwrap_or_default()));
             // insert old info into reverts.
-            reverts.entry(block_number).or_default().entry(address).or_default().0 = Some(old_info);
+            bundle_builder = bundle_builder.revert_account_info(
+                block_number,
+                address,
+                if let Some(old_info) = old_info {
+                    Some(Some(into_revm_acc(old_info)))
+                } else {
+                    Some(None)
+                },
+            );
         }
 
         // add storage changeset changes
         for (block_and_address, old_storage) in storage_changeset.into_iter().rev() {
             let BlockNumberAddress((block_number, address)) = block_and_address;
-            // get account state or insert from plain state.
-            let account_state = match state.entry(address) {
-                hash_map::Entry::Vacant(entry) => {
-                    let present_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((present_info, present_info, HashMap::new()))
-                }
-                hash_map::Entry::Occupied(entry) => entry.into_mut(),
-            };
-
-            // match storage.
-            match account_state.2.entry(old_storage.key) {
-                hash_map::Entry::Vacant(entry) => {
-                    let new_storage = plain_storage_cursor
-                        .seek_by_key_subkey(address, old_storage.key)?
-                        .filter(|storage| storage.key == old_storage.key)
-                        .unwrap_or_default();
-                    entry.insert((old_storage.value, new_storage.value));
-                }
-                hash_map::Entry::Occupied(mut entry) => {
-                    entry.get_mut().0 = old_storage.value;
-                }
-            };
+            let present_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
```

Comment:
Does this part seems okay, we now always override the present and original state, while previously we did it only if there is not account inside state `let account_state = match state.entry(address) {`

Reply:
ok, should have solved it now. I added a `Vec<Address>` that saves all seen addresses so that it's possible to filter out accounts already previously seen.